src/rgw/rgw_rest_s3.cc

```diff
@@ -6168,156 +6168,327 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
-  buff[i++] = char(7);
-  encode_short(&buff[i], uint16_t(strlen(header_value_str[EVENT])), i);
-  memcpy(&buff[i], header_value_str[EVENT], strlen(header_value_str[EVENT]));
-  i += strlen(header_value_str[EVENT]);
+  push_header(header_name_str[MESSAGE_TYPE], header_value_str[ERROR_TYPE]);
 
-  return i;
+  return m_buff_header.size();
 }
 
-int RGWSelectObj_ObjStore_S3::create_message(std::string &out_string, u_int32_t result_len, u_int32_t header_len)
+int aws_response_handler::create_message(u_int32_t header_len)
 {
-  //message description(AWS): 
+  //message description(AWS):
   //[total-byte-length:4][header-byte-length:4][crc:4][headers:variable-length][payload:variable-length][crc:4]
-  //s3select result is produced into m_result, the m_result is also the response-message, thus the attach headers and CRC 
+  //s3select result is produced into sql_result, the sql_result is also the response-message, thus the attach headers and CRC
   //are created later to the produced SQL result, and actually wrapping the payload.
 
+  auto push_encode_int = [&](u_int32_t s, int pos) {
+    u_int32_t x = htonl(s);
+    sql_result.replace(pos, sizeof(x), reinterpret_cast<char *>(&x), sizeof(x));
+  };
+
   u_int32_t total_byte_len = 0;
   u_int32_t preload_crc = 0;
   u_int32_t message_crc = 0;
-  int i = 0;
-  char * buff = out_string.data();
 
-  if(crc32 ==0) {
-    // the parameters are according to CRC-32 algorithm and its aligned with AWS-cli checksum
-    crc32 = std::unique_ptr<boost::crc_32_type>(new boost::crc_optimal<32, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, true, true>);
-  }
-
-  total_byte_len = result_len + 16;//the total is greater in 4 bytes than current size
+  total_byte_len = sql_result.size() + 4; //the total is greater in 4 bytes than current size
 
-  encode_int(&buff[i], total_byte_len, i);//store sizes at the beginning of the buffer
-  encode_int(&buff[i], header_len, i);
+  push_encode_int(total_byte_len, 0);
+  push_encode_int(header_len, 4);
 
   crc32->reset();
-  *crc32 = std::for_each( buff, buff + 8, *crc32 );//crc for starting 8 bytes
+  *crc32 = std::for_each(sql_result.data(), sql_result.data() + 8, *crc32); //crc for starting 8 bytes
   preload_crc = (*crc32)();
-  encode_int(&buff[i], preload_crc, i);
-
-  i += result_len;//advance to the end of payload.
+  push_encode_int(preload_crc, 8);
 
   crc32->reset();
-  *crc32 = std::for_each( buff, buff + i, *crc32 );//crc for payload + checksum
+  *crc32 = std::for_each(sql_result.begin(), sql_result.end(), *crc32); //crc for payload + checksum
   message_crc = (*crc32)();
-  char out_encode[4];
-  encode_int(out_encode, message_crc, i);
-  out_string.append(out_encode,sizeof(out_encode));
 
-  return i;
+  u_int32_t x = htonl(message_crc);
+  sql_result.append(reinterpret_cast<char *>(&x), sizeof(x));
+
+  return sql_result.size();
+}
+
+void aws_response_handler::init_response()
+{ //12 positions for header-crc
+  sql_result = "012345678901";
```

Comment:
the `aws_response_handler` is already initializing `sql_result`. this assignment is duplicated in a lot of other calls too